File: Gemi3/pyCAVI/_update_tau.py
import sys
from Gemi3.pyCAVI.util import find_s_keys,find_r_keys

def tau_loop(self):
    if self.verbose:
        print("Updating tau values.",file=sys.stdout)
    for tau_triple in self.hash_t.keys():
        indices=collect_all_keys(self,tau_triple)
        new_alpha = tau_alpha(self,indices)
        self.alpha[self.hash_t[tau_triple]] = new_alpha
        new_beta=tau_beta(self,indices)
        self.beta[self.hash_t[tau_triple]] = new_beta
        new_tau=new_alpha/new_beta
        if self.debug: print(f"Old tau {self.tau[self.hash_t[tau_triple]]}.", sys.stdout)
        self.tau[self.hash_t[tau_triple]] = new_tau
        if self.debug: print(f"New s {self.tau[self.hash_t[tau_triple]]}.", sys.stdout)

def tau_alpha(self,key_dict):
    return self.alpha[self.hash_t[key_dict["t"]]]+0.5

def tau_beta(self,key_dict):
    beta=self.beta[self.hash_t[key_dict["t"]]]
    D2=0.5 * self.D[self.hash_d[key_dict["t"]]]**2
    Dxgiygl=self.D[self.hash_d[key_dict["t"]]]*self.x[self.hash_x[key_dict["gi"]]]*self.y[self.hash_y[key_dict["gl"]]]
    Dxhjyhl=self.D[self.hash_d[key_dict["t"]]]*self.x[self.hash_x[key_dict["hj"]]]*self.y[self.hash_y[key_dict["hl"]]]
    Dxfkyfl=self.D[self.hash_d[key_dict["t"]]]*self.x[self.hash_x[key_dict["fk"]]]*self.y[self.hash_y[key_dict["fl"]]]
    Drs=self.D[self.hash_d[key_dict["t"]]]*self.r[self.hash_r[key_dict["r"]]]*self.s[self.hash_s[key_dict["s"]]]
    xgi2ygl2=self.xx[self.hash_x[key_dict["gi"]]]*self.yy[self.hash_y[key_dict["gl"]]]
    xgiyglxhjyhl=self.x[self.hash_x[key_dict["gi"]]]*self.y[self.hash_y[key_dict["gl"]]]*self.x[self.hash_x[key_dict["hj"]]]*self.y[self.hash_y[key_dict["hl"]]]
    xgiyglxfkyfl=self.x[self.hash_x[key_dict["gi"]]]*self.y[self.hash_y[key_dict["gl"]]]*self.x[self.hash_x[key_dict["fk"]]]*self.y[self.hash_y[key_dict["fl"]]]
    xgiyglrs=2*self.x[self.hash_x[key_dict["gi"]]]*self.y[self.hash_y[key_dict["gl"]]]*self.r[self.hash_r[key_dict["r"]]]*self.s[self.hash_s[key_dict["s"]]]
    xhj2yhl2=0.5*self.xx[self.hash_x[key_dict["hj"]]]*self.yy[self.hash_y[key_dict["hl"]]]
    xhjyhlxfkyfl=self.x[self.hash_x[key_dict["hj"]]]*self.y[self.hash_y[key_dict["hl"]]]*self.x[self.hash_x[key_dict["fk"]]]*self.y[self.hash_y[key_dict["fl"]]]
    xhjyhlrs=self.x[self.hash_x[key_dict["hj"]]]*self.y[self.hash_y[key_dict["hl"]]]*self.r[self.hash_r[key_dict["r"]]]*self.s[self.hash_s[key_dict["s"]]]
    r2s2=0.5*self.rr[self.hash_r[key_dict["r"]]]*self.ss[self.hash_s[key_dict["s"]]]
    out=beta+D2-Dxgiygl-Dxhjyhl-Dxfkyfl-Drs+xgi2ygl2+xgiyglxhjyhl+xgiyglxfkyfl+xgiyglrs+xhj2yhl2+xhjyhlxfkyfl+xhjyhlrs+r2s2
    return out


def collect_all_keys(self,tau_triple):
    gi,hj,fk,l = tau_triple.split("_")
    gl = self.guide_to_gene(gi)
    hl = self.guide_to_gene(hj)
    fl = self.guide_to_gene(fk)
    s = find_s_keys(self,gl,hl,fl,l)
    r = find_r_keys(self,gi,hj,fk)
    out={"t":tau_triple,"gi":gi,"hj":hj,"fk":fk,"l":l,
                    "gl":"_".join([gl,l]),
                    "hl":"_".join([hl,l]),
                    "fl":"_".join([hl,l]),
                    "r":r,
                    "s":s}
    return out

File: src/CAVI/update_tau.rs
use super::CAVI;
use std::collections::HashMap;
use itertools::Itertools;
use rayon::prelude::*;

impl<'a> CAVI<'a>{
    pub fn update_tau(&self) -> Result<(),Box<dyn std::error::Error>>{
        if self.verbose {
            println! {"Updating tau values"}
        }
        for gl in self.lookup.read().unwrap().get("tau").unwrap().keys(){
            if self.lookup.read().unwrap().get("tau").unwrap().get(gl).unwrap().len() != 1 {panic!("Wrong length of indices in update_tau.rs")}
            let new_alpha = tau_alpha(self,&self.lookup.read().unwrap().get("tau").unwrap().get(gl).unwrap()[0]);
            {*self.alpha.write().unwrap().get_mut(*self.hash_td.get(*gl).unwrap()).unwrap()= new_alpha}
            let new_beta = tau_beta(self,&self.lookup.read().unwrap().get("tau").unwrap().get(gl).unwrap()[0]);
            {*self.beta.write().unwrap().get_mut(*self.hash_td.get(*gl).unwrap()).unwrap()= new_beta}
            let new_tau = *self.alpha.read().unwrap().get(*self.hash_td.get(*gl).unwrap()).unwrap() / *self.beta.read().unwrap().get(*self.hash_td.get(*gl).unwrap()).unwrap();
            if self.debug {println!("Old tau {}.",*self.all_tau.read().unwrap().get(*self.hash_td.get(*gl).unwrap()).unwrap())}
            {*self.all_tau.write().unwrap().get_mut(*self.hash_td.get(*gl).unwrap()).unwrap()= new_tau}
            if self.debug {println!("New tau {}.",*self.all_tau.read().unwrap().get(*self.hash_td.get(*gl).unwrap()).unwrap())}
        }
        return Ok(());
    }
}


fn tau_alpha(obj : &CAVI,key_dict : &HashMap<&str,&str>) -> f64 {
    let out : f64 = *obj.alpha.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() + 0.5;
    return out;
    }

fn tau_beta(obj : &CAVI,key_dict : &HashMap<&str,&str>) -> f64 {
    let beta : f64 = *obj.beta.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap();
    let D2 : f64 = (0.5 * *obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap()).powf(2.0);
    let Dxgiygl : f64 = *obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap()).unwrap();
    let Dxhjyhl : f64 = *obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap();
    let Dxfkyfl : f64 = *obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("fk").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("fl").unwrap()).unwrap()).unwrap();
    let Drs : f64 = *obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() * *obj.all_r.read().unwrap().get(*obj.hash_r.get(*key_dict.get("r").unwrap()).unwrap()).unwrap() * *obj.all_s.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap();
    let xgi2ygl2 :f64 = *obj.all_xx.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap() * *obj.all_yy.read().unwrap().get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap()).unwrap();
    let xgiyglxhjyhl :f64 = *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap()).unwrap() * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap();
    let xgiyglxfkyfl : f64 =*obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap()).unwrap() * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("fk").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("fl").unwrap()).unwrap()).unwrap();
    let xgiyglrs :f64 = 2.0 * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("gl").unwrap()).unwrap()).unwrap() * *obj.all_r.read().unwrap().get(*obj.hash_r.get(*key_dict.get("r").unwrap()).unwrap()).unwrap() * *obj.all_s.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap();
    let xhj2yhl2 :f64 = 0.5 * *obj.all_xx.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_yy.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap();
    let xhjyhlxfkyfl :f64 = *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap() * *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("fk").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("fl").unwrap()).unwrap()).unwrap();
    let xhjyhlrs :f64 = *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap() * *obj.all_r.read().unwrap().get(*obj.hash_r.get(*key_dict.get("r").unwrap()).unwrap()).unwrap() * *obj.all_s.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap();
    let r2s2 :f64 = 0.5 * *obj.all_rr.read().unwrap().get(*obj.hash_r.get(*key_dict.get("r").unwrap()).unwrap()).unwrap() * *obj.all_ss.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap();
    let out : f64 = beta+D2-Dxgiygl-Dxhjyhl-Dxfkyfl-Drs+xgi2ygl2+xgiyglxhjyhl+xgiyglxfkyfl+xgiyglrs+xhj2yhl2+xhjyhlxfkyfl+xhjyhlrs+r2s2;
    return out;
    }
